### PR TITLE
[KOGITO-7754] create Knative eventing resources

### PR DESCRIPTION
**Pull request by RichardW98**
create Knative resources:
1. create sinkbinding and trigger  if `spec.flow.event` is provided
2. tests

### CODE REVIEW

**Review by ricardozanini — COMMENTED** [review]
Great work! I think we are on the right path. I just added a few comments that we can make the implementation simpler and a note about triggers. Right on! Many thanks!

File: controllers/profiles/common/object_creators.go [hunk 1]
```diff
@@ -197,6 +211,93 @@ func ServiceCreator(workflow *operatorapi.SonataFlow) (client.Object, error) {
...
+		// filter out produce events
+		if event.Kind == cncfmodel.EventKindProduced {
+			continue
+		}
+		sink := workflow.Spec.Sink
+		if sink == nil {
+			return nil, errors.New("Sink is not provided to create SinkBinding!")
+		} else if sink.Ref.Kind != "Broker" || sink.Ref.APIVersion != "eventing.knative.dev/v1" {
+			return nil, errors.New("Trigger source must be a broker!")
+		}
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/common/object_creators.go:271]: yes,  I think we forgot to discuss this in the meeting. It was my bad that I thought trigger can use non-broker as source. Let's do this with defaults for this pr and create a follow up one to allow user customization.

File: controllers/profiles/common/mutate_visitors.go [hunk 2]
```diff
@@ -103,6 +106,41 @@ func ServiceMutateVisitor(workflow *operatorapi.SonataFlow) MutateVisitor {
...
+			if kubeutil.IsObjectNew(object) {
+				return nil
+			}
+			originals, err := TriggersCreator(workflow)
+			if err != nil {
+				return err
+			}
+			for _, original := range originals {
+				if original.GetName() == object.GetName() {
+					return mergo.Merge(&object.(*v12.Trigger).Spec, original.(*v12.Trigger).Spec, mergo.WithOverride)
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/common/mutate_visitors.go:136]: actually I remember the admission webhook will block the `broker` update

[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/common/mutate_visitors.go:136]: Screenshot 2024-01-11 at 9 21 31 AM

[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/common/mutate_visitors.go:136]: is it ok to only reconcile the generated fields? e.g. event-type, event-filter

File: controllers/profiles/prod/deployment_handler.go [hunk 3]
```diff
@@ -77,6 +79,50 @@ func (d *deploymentReconciler) reconcileWithBuiltImage(ctx context.Context, work
...
+
+		if sinkBindingResult != controllerutil.OperationResultNone {
+			// create, update should change the status
+			workflow.Status.Manager().MarkFalse(api.KnativeResourcesConditionType, api.WaitingForKnativeReason, "")
+		} else {
+			for _, trigger := range triggers {
+				if trigger.Result != controllerutil.OperationResultNone {
+					workflow.Status.Manager().MarkFalse(api.KnativeResourcesConditionType, api.WaitingForKnativeReason, "")
+					break
+				}
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/prod/deployment_handler.go:121]: I will move this to common.FollowWorkflowKnativeState which manages the knative ensurers as well

File: controllers/profiles/common/mutate_visitors.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ricardozanini) [reply controllers/profiles/common/mutate_visitors.go]: Oh I see! Yes, sure! Let's only reconcile the generated fields. This generated trigger can be used as a reference to users to create their own. It's enough for a first iteration.

File: controllers/profiles/common/object_creators.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ricardozanini) [reply controllers/profiles/common/object_creators.go:270]: Can you create this issue?

File: controllers/profiles/common/knative.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+			"no need to create Knative eventing resources since no event definition is found")
+	} else if workflow.Spec.Sink == nil {
+		// mark false if spec.sink is not found
+		workflow.Status.Manager().MarkFalse(api.KnativeResourcesConditionType,
+			api.KnativeFailureReason,
+			"Spec.Sink is not provided")
+	} else if f.Catalog != nil && f.Catalog.(*discovery.SonataFlowServiceCatalog).KnativeCatalog == nil {
+		workflow.Status.Manager().MarkFalse(api.KnativeResourcesConditionType,
+			api.KnativeFailureReason,
+			"Knative Client is not installed")
```
Comment (ricardozanini) [inline controllers/profiles/common/knative.go:70]:
```suggestion
			"Knative Eventing is not installed")
```

File: controllers/profiles/common/object_creators.go [hunk 5]
```diff
@@ -197,6 +213,90 @@ func ServiceCreator(workflow *operatorapi.SonataFlow) (client.Object, error) {
...
+		// filter out produce events
+		if event.Kind == cncfmodel.EventKindProduced {
+			continue
+		}
+		sink := workflow.Spec.Sink
+
+		//TODO: user should have the flexibility to config trigger source broker
+		if sink.Ref.Kind != "Broker" || sink.Ref.APIVersion != "eventing.knative.dev/v1" {
+			return nil, errors.New("Trigger source must be a broker!")
+		}
```
Comment (ricardozanini) [inline controllers/profiles/common/object_creators.go:270]: You can remove this code and hardcode to the trigger's broker to `default` for now.

File: controllers/profiles/profile.go [hunk 6]
```diff
@@ -71,6 +71,8 @@ type ReconciliationState interface {
 	Do(ctx context.Context, workflow *operatorapi.SonataFlow) (ctrl.Result, []client.Object, error)
 	// PostReconcile performs the actions to perform after the reconciliation that are not mandatory
 	PostReconcile(ctx context.Context, workflow *operatorapi.SonataFlow) error
+	// CanReconcileNext check if next state can be reconciled
+	CanReconcileNext() bool
```
Comment (ricardozanini) [inline controllers/profiles/profile.go:75]: Why can't we wait for the Knative reconciliation state to finish to proceed to others? Not sure if we have to change this API for this feature. I'd rather keep it simpler. I think we can keep it the way it was and each reconciliation cycle be deterministic on the state it supposed to run.

Actually, not sure why we need a Knative state. That can happen within the deployment handler. Not even the Knative conditions we need, sounds off.

We can instead post the Knative objects status to the events API. The more conditions and states we add, the more complex it gets. Now adding a new API to the core reconcile might make things more difficult to maintain.

Not sure if I was clear in the last review, but within the existing state where we create everything, we add a new `KnativeEventingHandler` as we have here: 
https://github.com/apache/incubator-kie-kogito-serverless-operator/blob/main/controllers/profiles/prod/deployment_handler.go#L50

So the code from the Knative state goes to a handler like this in the `knative.go` file. You reuse the whole states we already have in both profiles.

For example, in dev mode you add here:

https://github.com/apache/incubator-kie-kogito-serverless-operator/blob/main/controllers/profiles/dev/states_dev.go#L111C28-L111C28

Something like:

```go
knative_objs := common.KnativeHandler().Ensure()
```

In prod, you use the `deployment_handler.go`:
https://github.com/apache/incubator-kie-kogito-serverless-operator/blob/main/controllers/profiles/prod/deployment_handler.go#L78

**Review by ricardozanini — COMMENTED** [review]
Hi! I think we need a few movements here and there, but I can see that the feature is already there. Great work!

File: controllers/profiles/profile.go (hunk 6 above)
Comment (RichardW98) [reply controllers/profiles/profile.go:75]: sorry I misunderstood your last review. this logic here indeed adds much more complexity that we don't need as sonataflow opr. Users should be at their own obligation for the reconcile of these objs too.  I will remove the Knative state and Knative status condition,  and move the ensurer to the handlers 🙂

**Comment by RichardW98** [comment]
the extra knative status and state is removed. now user can edit the knative objs based on their needs. please review again

File: controllers/profiles/common/object_creators.go [hunk 7]
```diff
@@ -197,6 +211,86 @@ func ServiceCreator(workflow *operatorapi.SonataFlow) (client.Object, error) {
...
+
+		// construct eventingv1.Trigger
+		trigger := &eventingv1.Trigger{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      strings.ToLower(fmt.Sprintf("%s-%s-trigger", workflow.Name, event.Name)),
+				Namespace: workflow.Namespace,
+				Labels:    lbl,
+			},
+			Spec: eventingv1.TriggerSpec{
+				Broker: sink.Ref.Name,
```
Comment (ricardozanini) [inline controllers/profiles/common/object_creators.go:273]:
```suggestion
				Broker: "default",
```
We can have a const `knativeEventingBrokerDefault = "default"`

**Review by ricardozanini — COMMENTED** [review]
One small last thing :P

**Comment by ricardozanini** [comment]
Just verify the CI please.

@domhanak can you start the verification? 🙏

File: controllers/profiles/common/object_creators.go (hunk 7 above)
Comment (RichardW98) [reply controllers/profiles/common/object_creators.go:273]: oh yes make it hardcode haha got it

**Review by wmedvede — COMMENTED** [review]
Hi @RichardW98 good job! 
I have added some comments, but couldn't finish all the review, on Monday I'll finish the remaining files.

Many thanks!

File: controllers/profiles/common/ensurer.go [hunk 8]
```diff
@@ -66,22 +67,10 @@ func (d *defaultObjectEnsurer) Ensure(ctx context.Context, workflow *operatorapi
 	result := controllerutil.OperationResultNone
 
 	object, err := d.creator(workflow)
-	if err != nil {
+	if err != nil || object == nil {
 		return nil, result, err
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/common/ensurer.go:71]: the reason I put return here for 'object==nil' is cuz if the it is nil, then we don't need to ensure it and also don't need to append it to objs in state

File: controllers/profiles/common/object_creators.go [hunk 9]
```diff
@@ -197,6 +211,85 @@ func ServiceCreator(workflow *operatorapi.SonataFlow) (client.Object, error) {
...
+	if workflow.Spec.Sink == nil || !workflowdef.ContainsEventKind(workflow, cncfmodel.EventKindProduced) {
+		return nil, nil
+	}
+
+	sink := workflow.Spec.Sink
+
+	// subject must be deployment to inject K_SINK, service won't work
+	sinkBinding := &sourcesv1.SinkBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      strings.ToLower(fmt.Sprintf("sb-%s", workflow.Name)),
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/common/object_creators.go:229]: yea totally agree on the naming convention optimization

File: controllers/discovery/discovery.go [hunk 10]
```diff
@@ -80,35 +80,34 @@ type ServiceCatalog interface {
 	Query(ctx context.Context, uri ResourceUri, outputFormat string) (string, error)
 }
 
-type sonataFlowServiceCatalog struct {
+type SonataFlowServiceCatalog struct {
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/discovery/discovery.go:83]: it's for checking if knative client is available in https://github.com/apache/incubator-kie-kogito-serverless-operator/pull/350/files#diff-f4169987ade59d934bdc042b4e7e6147bc4b8848bcda924eab7425e352763200R55

File: controllers/discovery/discovery.go [hunk 11]
```diff
@@ -80,35 +80,34 @@ type ServiceCatalog interface {
...
 	kubernetesCatalog ServiceCatalog
-	knativeCatalog    ServiceCatalog
+	KnativeCatalog    ServiceCatalog
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/discovery/discovery.go:85]: same reason as above

File: controllers/discovery/discovery.go [hunk 12]
```diff
@@ -80,35 +80,34 @@ type ServiceCatalog interface {
...
 		openshiftCatalog:  newOpenShiftServiceCatalog(openShiftDiscoveryClient),
 	}
 }
 
 func NewServiceCatalogForConfig(cli client.Client, cfg *rest.Config) ServiceCatalog {
-	return &sonataFlowServiceCatalog{
+	return &SonataFlowServiceCatalog{
 		kubernetesCatalog: newK8SServiceCatalog(cli),
-		knativeCatalog:    newKnServiceCatalogForConfig(cfg),
-		openshiftCatalog:  newOpenShiftServiceCatalogForClientAndConfig(cli, cfg),
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/discovery/discovery.go:102]: yea.. the removal of openshiftCatalog is a mistake. Thank you for catching it!

[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/discovery/discovery.go:102]: the change on capitalizing first char of name is just for checking knative availability.

File: controllers/profiles/common/properties/properties.go [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import (
+	operatorapi "github.com/apache/incubator-kie-kogito-serverless-operator/api/v1alpha08"
+	"github.com/apache/incubator-kie-kogito-serverless-operator/controllers/profiles/common/constants"
+	"github.com/apache/incubator-kie-kogito-serverless-operator/controllers/workflowdef"
+	"github.com/magiconair/properties"
+	cncfmodel "github.com/serverlessworkflow/sdk-go/v2/model"
+)
+
+// GenerateKnativeEventingWorkflowProperties returns the set of application properties required for the workflow to produce or consume
+// Knative Events. For the calculation this function considers if the Job Service is present in the
```
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/profiles/common/properties/properties.go:26]: I will remove the wrong descriptions lol

File: controllers/discovery/discovery.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (wmedvede) [reply controllers/discovery/discovery.go:102]: I think you can add a function here https://github.com/apache/incubator-kie-kogito-serverless-operator/blob/f0d9068f9fde4ab2476b76c1c75d5567e1428fc2/controllers/knative/knative.go  for that purpose, and rollback the changes in sonataFlowServiceCatalog.

The ServiceCatalog implementation we have, far from perfect, is intended to use the interface and not relay on the particular implementation and it's internal fields.

**Comment by wmedvede** [comment]
Hi @RichardW98 , let me add some last comments regarding some missing tweaks that we need, and which is totally fine since you are not yet familiar with all the nitpicks we need to complete the configuration.

At the workflow side, you are already doing some iteration [](https://github.com/apache/incubator-kie-kogito-serverless-operator/blob/df36511f9a5344611eff714d9d510200813d4784/controllers/profiles/common/object_creators.go#L220) to determine if the workflow produces events, etc., and thus, to create the **SinkBinding**, which is good.

However, there are more situations where the workflow might need to send cloud events, even in cases where we don't have any cncfmodel.EventKindProduced at all. That is the case of workflows that use **timeouts** (might produce other events or not), since the implementation of these **timeouts** involves sending cloud events to the Job service.

**Events production detection:**

So, additionally to this cncfmodel.EventKindProduced you are doing, we must check if the workflow has timeouts. And right now, with the current SW timeouts support we have, I think we must focus on inspecting the following situations:

1. At the top-level workflow definition, we must check if the workflow has configured any of the following timeouts: **workflowExecTimeout** or **eventTimeout**

https://github.com/serverlessworkflow/specification/blob/main/specification.md#workflow-timeouts

**example here:**  https://github.com/apache/incubator-kie-kogito-examples/blob/a60d2f0b1ecfdab0758da74bc799adf53a83f5d5/serverless-workflow-examples/serverless-workflow-timeouts-showcase-extended/src/main/resources/workflow-timeouts.sw.json#L8


2. If the workflow has any a **sleep** state as part of its definition

https://github.com/serverlessworkflow/specification/blob/main/specification.md#sleep-state

3. If the workflow has any **callback** state with a configured **eventTimeout** as part of its definition

https://github.com/serverlessworkflow/specification/blob/main/specification.md#callback-state

**example here:**  https://github.com/apache/incubator-kie-kogito-examples/blob/2a28c6ae545622c95dfa3ca31b9df525b4c56e87/serverless-workflow-examples/serverless-workflow-timeouts-showcase-extended/src/main/resources/callback-state-timeouts.sw.json#L53

4. If the workflow has any **event** state with a configured **eventTimeout** as part of its definition

https://github.com/serverlessworkflow/specification/blob/main/specification.md#event-state

**example here:**  https://github.com/apache/incubator-kie-kogito-examples/blob/2a28c6ae545622c95dfa3ca31b9df525b4c56e87/serverless-workflow-examples/serverless-workflow-timeouts-showcase-extended/src/main/resources/event-state-timeouts.sw.json#L89

5) If the workflow has any **switch** state with a configured **eventTimeout** as part of its definition

https://github.com/serverlessworkflow/specification/blob/main/specification.md#switch-state

**example here:**  https://github.com/apache/incubator-kie-kogito-examples/blob/2a28c6ae545622c95dfa3ca31b9df525b4c56e87/serverless-workflow-examples/serverless-workflow-timeouts-showcase-extended/src/main/resources/switch-state-timeouts.sw.json#L60


**K_SINK health check:**

When we have knative enabled, and we move to use the knative eventing, etc., we use the K_SINK env var, etc.
In that cases, at the workflows runtime side, we incorporate a heath check on that variable, to ensure we have value there. This is good, because the runtime will rely on the K_SINK value to deliver the events.

Right now, we are intentionally disabling that check, because we are not using knative eventing, so we don't want the workflow health check to fail because of this.

Now, as part of this PR, if the user decides to use knative eventing, and we determine that the SinkBinding must be created, we must enable it.

To manage that health check we use the following property org.kie.kogito.addons.knative.eventing.health-enabled, which is right now defaulted to false as an immutable property.

see here:   https://github.com/apache/incubator-kie-kogito-serverless-operator/blob/df36511f9a5344611eff714d9d510200813d4784/controllers/profiles/common/properties/application.go#L45

File: controllers/discovery/discovery.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (RichardW98) [reply controllers/discovery/discovery.go]: I just changed to use GetKnativeAvailability() for the check

**Comment by ricardozanini** [comment]
@RichardW98 please check the files generation:

```shell
make generate-all
make vet fmt
```

**Comment by RichardW98** [comment]
hi Ricardo, I just fixed the uni tests, please help check again

**Review by ricardozanini — APPROVED** [review]
@wmedvede @RichardW98 do we have the follow up issues? If so, can you please link them here? 🙏

**Comment by RichardW98** [comment]
yes, I just created some followup issues, please find them under epic: https://issues.redhat.com/browse/KOGITO-9812 and the list is here:

[KOGITO-10031](https://issues.redhat.com/browse/KOGITO-10031)	allow user customization on generated knative eventing resources
[KOGITO-10032](https://issues.redhat.com/browse/KOGITO-10032)	create knative resources for DataIndex and JobService
[KOGITO-10033](https://issues.redhat.com/browse/KOGITO-10033)	Support Multiple Sinks
[KOGITO-10034](https://issues.redhat.com/browse/KOGITO-10034)	creation of sinks in Sonataflow
[KOGITO-10035](https://issues.redhat.com/browse/KOGITO-10035)	update the knative properties to immutable env var in deployment

**Comment by ricardozanini** [comment]
@RichardW98 Thank you, but we are not using JIRA anymore. We used it for this one 'cause it was already opened. We should have issues within this repo and assign them to the milestone. Sorry about that, I'll open the issues here.

**Review by ricardozanini — APPROVED** [review]

**Comment by ricardozanini** [comment]
@RichardW98 can you take a look at the E2E failure?

File: controllers/profiles/common/object_creators.go [hunk 14]
```diff
@@ -202,6 +216,85 @@ func ServiceCreator(workflow *operatorapi.SonataFlow) (client.Object, error) {
 	return service, nil
 }
 
+// SinkBindingCreator is an ObjectsCreator for SinkBinding.
+// It will create v1.SinkBinding based on events defined in workflow.
+func SinkBindingCreator(workflow *operatorapi.SonataFlow) (client.Object, error) {
+	lbl := workflowproj.GetDefaultLabels(workflow)
+
+	// skip if no produced event is found
+	if workflow.Spec.Sink == nil || !workflowdef.ContainsEventKind(workflow, cncfmodel.EventKindProduced) {
```
Comment (wmedvede) [inline controllers/profiles/common/object_creators.go:225]: The cases we chatted, regarding the "timeouts" and the situations where a workflow might produce no regular events, but can still send cloud events to the jobs service. Are you going to implement in a complementary PR?

File: controllers/profiles/common/properties/properties.go [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+// Knative Events.
+// Never nil.
+func generateKnativeEventingWorkflowProperties(workflow *operatorapi.SonataFlow) (*properties.Properties, error) {
+	props := properties.NewProperties()
+	if workflow == nil || workflow.Spec.Sink == nil {
+		props.Set(constants.KnativeHealthEnabled, "false")
+		return props, nil
+	}
+	// verify ${K_SINK}
+	props.Set(constants.KnativeHealthEnabled, "true")
```
Comment (wmedvede) [inline controllers/profiles/common/properties/properties.go:35]: This property must be enabled in cases that we are sure that the SinkBinding will be created. If we set it to true here, but then, we the SinkBinding is not created, we'll have issues.

File: controllers/profiles/common/object_creators.go (hunk 14 above)
Comment (RichardW98) [reply controllers/profiles/common/object_creators.go:225]: yes, I will do it in PR for job services: https://github.com/apache/incubator-kie-kogito-serverless-operator/issues/368. this is gonna be my next pr

File: controllers/profiles/common/properties/properties.go (hunk 15 above)
Comment (RichardW98) [reply controllers/profiles/common/properties/properties.go:35]: yes, this function will set it to false if Spec.Sink is nil, which means there'll be no sinkbinding to be create: https://github.com/apache/incubator-kie-kogito-serverless-operator/issues/368.

File: controllers/profiles/common/object_creators.go (hunk 14 above)
Comment (wmedvede) [reply controllers/profiles/common/object_creators.go:225]: btw, I'm implementing this calculation already in my upcoming PR. I'll let you know when merged.

Comment (RichardW98) [reply controllers/profiles/common/object_creators.go:225]: great! Thank you

**Comment by ricardozanini** [comment]
@RichardW98 we will put this on hold until we stabilize the features for the next two weeks.

**Review by wmedvede — APPROVED** [review]

**Comment by ricardozanini** [comment]
@RichardW98 can you please rebase?

**Pull request merged**